### PR TITLE
Google Cloud Functions Codestart

### PR DESCRIPTION
**Pull request by loicmathieu**

### CODE REVIEW

**Comment by loicmathieu** [comment]
@ia3andy I finally had some time to play with codestarts, this is easy I must say :)

This is the Google Cloud Functions codestart, I generated it with jbang to check that it worked. Do I need to create a test somewhere ?

After generating the example project, there is still two thing that are not needed for Google Cloud Functions:
- Dockerfile: is there a way to avoid generating them ?
- Existing README: it explains how to run and package the application but with Google Cloud Functions the standard way to run and package the application didn't works

**Comment by ia3andy** [comment]
Thanks!! I will have look today, I was in PTO..

**Comment by ia3andy** [comment]
> @ia3andy I finally had some time to play with codestarts, this is easy I must say :)
> 
> This is the Google Cloud Functions codestart, I generated it with jbang to check that it worked. Do I need to create a test somewhere ?

It will be tested alongside the others (generated and built) in `QuarkusCodestartRunIT`

> 
> After generating the example project, there is still two thing that are not needed for Google Cloud Functions:
> 
> * Dockerfile: is there a way to avoid generating them ?

Well there is a `noDockerfiles` option in `CreateProject`, but it's not plugged yet to the CLI and Quarkus Maven plugin. So either we could make it automatic (with a flag in codestart.yml) or manual with the option, I am wondering why the dockerfiles are not needed for this?

> * Existing README: it explains how to run and package the application but with Google Cloud Functions the standard way to run and package the application didn't works

As you can select it alongside other examples which could work the standard way, the convention is to add a warning in the README.md fragment like for `Funqy Amazon Lambda Binding`:
```
# Funqy Amazon Lambda Binding

> :warning: **INCOMPATIBLE WITH DEV MODE**: Funqy Amazon Lambda Binding is not compatible with dev mode yet!

```

**Comment by ia3andy** [comment]
Ah sorry I missed that it was a `singleton-example` :)

- It will be automatically tested but alone..
- for the README.md, for now the warning is fine. But maybe we should create an issue to find a proper solution as we have the guaranty that we have only one example for `singleton`

**Review by ia3andy — CHANGES_REQUESTED** [review]
Cool! 

I've made a few suggestion in my review, also, would it be relevant to generate an example test for this codestart?

**Comment by loicmathieu** [comment]
@ia3andy field #14126 and #14125 for codestart enhancement for functions.

**Comment by loicmathieu** [comment]
@ia3andy I applied your suggestion

Regarding 

> would it be relevant to generate an example test for this codestart?

There is no specific testing support so you cannot just launch the application for testing (it's on my TODO list). So testing should be done manually or via injection of the function in a test case. Injection of a function in a test case would need mocking for the function parameter which can be difficult for HttpRequest and HttpResponse.
So there is no easy way yet to test Google Cloud functions :(

File: devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/google-cloud-functions-example/codestart.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+name: google-cloud-functions-example
+ref: google-cloud-functions
```
Comment (ia3andy) [inline devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/google-cloud-functions-example/codestart.yml:2]: Sorry I missed that you didn't add the extension binding in the descriptor metadata:
https://github.com/quarkusio/quarkus/blob/master/extensions/google-cloud-functions/runtime/src/main/resources/META-INF/quarkus-extension.yaml
```yaml
codestart: google-cloud-functions
```

(This is what trigger the codestart activation when the extension is selected)

I am not sure if the binding should be on `google-cloud-functions` or `google-cloud-functions-http` (or both)

**Review by ia3andy — CHANGES_REQUESTED** [review]

**Review by ia3andy — APPROVED** [review]
Thanks @loicmathieu :)

File: devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/google-cloud-functions-example/codestart.yml (hunk 1 above)
Comment (loicmathieu) [reply devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/google-cloud-functions-example/codestart.yml:2]: It's `google-cloud-functions` only, I'll provide a separate PR for HTTP binding and for funqy.
I added it in the corresponing `quarkus-extension.yaml` file.

By the way, this step is missing from the documentation: https://github.com/quarkusio/quarkus/tree/master/independent-projects/tools/codestarts

Comment (ia3andy) [reply devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/google-cloud-functions-example/codestart.yml:2]: Weird the adoc rendering is not working, it was working fine a few weeks ago..

@loicmathieu it is :) look at the "When done (to bind the extension codestart to the actual extension)" section

Comment (loicmathieu) [reply devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/google-cloud-functions-example/codestart.yml:2]: Ah OK, I read too quickly.

This deserve a little code snippet from an extension :)

**Pull request merged**